rust/connlib/tunnel/src/lib.rs

```diff
@@ -122,6 +123,11 @@ where
             Poll::Pending => {}
         }
 
+        // After any state change, check what the new timeout is and reset it if necessary.
+        if self.connections_state.poll_timeout(cx).is_ready() {
+            cx.waker().wake_by_ref()
+        }
+
         Poll::Pending
```

Comment:
Just because I'm still getting used to the `poll_` style, do I understand this right?
- We call `poll_timeout` on this other state object and pass it our context, so if it needs to register a wakeup, it may do that
- If the connection state has something ready, we do a `wake_by_ref` to tell the driver / runtime, "Immediately call me again, there might be more work to do"
- We return `Pending` because whatever `connections_state` needs to do is already handled above here and we don't want to duplicate calling it in these other two blocks. But probably the runtime will immediately wake up our task again and then we'll take care of whatever timed out for `connections_state`.
- In general if we return `Ready` or have a subtask (sub state machine?) that's ready, the runtime should keep polling us until we return `Pending` and also neither us nor our subtasks register an immediate wakeup? e.g. if there were no timeouts somehow, we would be polled until everything was done, and then only wake up when a new packet comes in?
- I'm guessing `connections_state` manages several connections, so internally its `poll_timeout` checks on something like a timer heap, so if we have 5 connections, it just tells us which timeout will happen earliest.

Reply:
Yes you've got it all right :)

The runtime will end up polling a future which is actually never `Ready` because that would shut it down.

Reply:
At the moment, the eventloops are still a bit fragile in regards to the polling rules but I am hoping to refactor that and make it crystal clear and easy to follow!

Reply:
> * All this runs in a single thread because the runtime overhead of locks and the dev overhead of multi-threading (having to deal with everything being `&self` and mixing sync and async) is not worth it, since we are not CPU-bound. If we needed an outrageous amount of bandwidth, like 1 Tbps, or if cryptography was more expensive, one CPU core would not suffice, but for our case of ballpark 1 Gbps, there's no need to use extra cores?

That is the bet yeah and so far it seems to turn out okay :)

There is even more optimisation potential like io-uring to further reduce CPU overhead. It would be interesting to know, what link speeds you need for this design fails.